feat: (IAC-895) Add Support to Toggle the Google Cloud Managed Service for Prometheus

jarpat opened the pull request:
### Changes

After March 15, 2023 all newly created Google Kubernetes Engine standard clusters will have [Managed Service for Prometheus](https://cloud.google.com/stackdriver/docs/managed-prometheus) in-cluster components deployed by default.

This default enabled feature from Google does add a small amount of cost. See https://cloud.google.com/stackdriver/pricing#mgd-prometheus-pricing-summary

In order to support users who do want this service in their cluster by default we are adding the `enable_managed_prometheus` to control the deployment of the  "Managed Service for Prometheus" in their clusters. By default we are setting this value to false.

Note: Google has still not enabled this feature by default, but adding the flag as a feature in case users want to enable this themselves.

### Tests

| Scenario | Provider | Kubernetes Version     | Deployment Method | `enable_managed_prometheus` | Notes                              |
|----------|----------|------------------------|-------------------|---------------------------|------------------------------------|
| 1        | GCP      | 1.26 (v1.26.4-gke.500) | Docker            | true                      | GMP was present in the cluster     |
| 2        | GCP      | 1.26 (v1.26.4-gke.500  | Docker            | unset (default false)     | GMP was NOT present in the cluster |


Additional tests after review comments:

### Tests

| Scenario | Provider | Kubernetes Version     | Deployment Method | enable_managed_prometheus | create_gke_monitoring_service | gke_monitoring_service               | gke_monitoring_enabled_components     | Notes                                                      |
|----------|----------|------------------------|-------------------|---------------------------|-------------------------------|--------------------------------------|---------------------------------------|------------------------------------------------------------|
| 3        | GCP      | 1.26 (v1.26.4-gke.500) | Docker            | true                      | true                          | monitoring.googleapis.com/kubernetes | unset (default ["SYSTEM_COMPONENTS"]) | GMP was present in the cluster & gke-metrics-agent present |
| 4        | GCP      | 1.26 (v1.26.4-gke.500) | Docker            | true                      | false                         | unset (default none n/a)             | unset n/a                             | GMP was present in the cluster gke-metrics-agent absent    |

thpang commented on docs/CONFIG-VARS.md:
````diff
@@ -293,3 +294,9 @@ postgres_servers = {
 }
 ```
+
+## Monitoring
````
We call this monitoring; however, this was already part of the code base in `create_gke_monitoring_service` what/how will we distinguish these items and keep the confusion down with our customers? I know AWS has a prometheus offering and Azure has their version in preview, Is prometheus a variant of monitoring? and/or a solution being offered as its the industry go to?

jarpat replied:
Summary of : https://cloud.google.com/stackdriver/docs/solutions/gke

The main difference is that `gke_monitoring_service` (Cloud Monitoring service add-on) is used to to monitor both the the system and workload components of the cluster. However I just found out that as of 1.24 workload monitoring is deprecated and Google recommends that the Google Cloud Managed Service for Prometheus is used monitor and alert on your workloads. Quote on bottom.

Even through the monitoring is split up between the GKE add on and Prometheus all the data will still flow into the GCP monitoring service here https://console.cloud.google.com/monitoring

Another behavior that I saw is that with `enable_managed_prometheus` and `gke_monitoring_service` both enabled the user needs to explicitly specify which components the `gke_monitoring_service` (the add-on) will monitor, this requires one additional variable `gke_monitoring_enabled_components` that I am in process of adding.

I saw this in the GKE details page

>Cloud Monitoring is a Google Kubernetes Engine (GKE) addon that collects metrics emitted by your applications and by GKE infrastructure. Learn more 

>System monitoring includes over 40 platform metrics such as CPU, memory, and storage for essential system components. Cloud Monitoring does not charge for the ingestion of system metrics. Learn more 

>Workload monitoring is deprecated. Workload monitoring is removed in GKE 1.24 and later. Workload monitoring is replaced by Google Cloud Managed Service for Prometheus, which is Google's recommended way to monitor Kubernetes applications by using Cloud Monitoring. Learn more

riragh commented on variables.tf:
```diff
@@ -389,6 +389,12 @@ variable "gke_monitoring_service" {
+variable "enable_managed_prometheus" {
+  type        = bool
+  description = "Enable Google Cloud Managed Service for Prometheus for your cluster"
+  default     = "false"
```
false shouldn't be quoted as it's a Boolean value.

jarpat replied:
fixed in latest push

jarpat commented:
Added documentation for the previously undocumented `gke_monitoring_*` flags and added ther now required `gke_monitoring_enabled_components` which I am working on testing. I'll requests reviews after testing is complete.
